Adding cgroup constrain for containerd and memlogd

adarsh-zededa opened the pull request:
Added a cgroup for containerd and memlogd

Restructured cgroup hierarchy as below whose limits can be changed using respective variables in grub.cfg.

eve (parent, can be modified using `hv_dom0_mem_settings` and `hv_dom0_cpu_settings`)
- services (cgroup parent for all eve services, can be modified using `hv_eve_mem_settings` and `hv_eve_cpu_settings`)
-- rsyslogd 
-- ntpd 
-- sshd 
-- wwan 
-- wlan 
-- lisp 
-- guacd 
-- pillar 
-- vtpm 
-- watchdog 
-- xen-tools
- containerd (can be modified using `hv_ctrd_mem_settings` and `hv_ctrd_cpu_settings`)
- memlogd

Added a config var in grub to manipulate watchdog timer. 

NOTE: For testing purpose removed LISP service.

Signed-off-by: adarsh-zededa <adarsh@zededa.com>

deitch commented:
This is starting to look really nice. Specifically, it is starting to have a "design feel" to it that makes it easier to grasp and understand.

I do have a few minor comments.

* This needs a rebase since Roman's PR on `config.toml` went through
* Why are we calling the cgroups `/eve/services/eve-*`? It seems the `eve-` is redundant. If there is a good reason, let us keep it, but if not, then it probably would be cleaner to just be `/eve/services/volumemgr` or similar. Yes, I am aware it was there before, but I only realized it now :-)
* Why is memlogd under `/eve/memlogd` but everything else is under `/eve/services/<name>`?
* A nit, but I think it is easier to understand putting the `sh -c <cmd>` in `/etc/init.d/002-logging` is just wrapped in single `'` rather than `""`, so we won't have to escape out the `$$`, i.e.: `sh -c 'echo $$ > /sys/fs/cgroup/...`

And one larger comment: we have to have this documented. The fact that there are defaults restricting resource usage, and they can be overridden, cannot just be in the code or config files. There _has_ to be a bright, red, bold document that this exists, the impact and how to tweak them.

adarsh-zededa commented:
> * This needs a rebase since Roman's PR on `config.toml` went through

Done.

> * Why are we calling the cgroups `/eve/services/eve-*`? It seems the `eve-` is redundant. If there is a good reason, let us keep it, but if not, then it probably would be cleaner to just be `/eve/services/volumemgr` or similar. Yes, I am aware it was there before, but I only realized it now :-)

Removed "eve-"

> * Why is memlogd under `/eve/memlogd` but everything else is under `/eve/services/<name>`?

The idea was to keep the services (present under linuxkit config's _services_) under a separate cgroup.

> * A nit, but I think it is easier to understand putting the `sh -c <cmd>` in `/etc/init.d/002-logging` is just wrapped in single `'` rather than `""`, so we won't have to escape out the `$$`, i.e.: `sh -c 'echo $$ > /sys/fs/cgroup/...`

Changed.

> And one larger comment: we have to have this documented. The fact that there are defaults restricting resource usage, and they can be overridden, cannot just be in the code or config files. There _has_ to be a bright, red, bold document that this exists, the impact and how to tweak them.

I'll update this PR with the doc changes.

deitch commented:
> The idea was to keep the services (present under linuxkit config's services) under a separate cgroup.

That makes sense. Let's include that pattern in the doc, so people understand why, and where to put future services.

eriknordmark commented on pkg/watchdog/init.sh:
```diff
@@ -36,7 +43,7 @@ run_watchdog() {
       (cat /etc/watchdog.conf.seed
        find /run/watchdog -type f | sed -e 's#/run/watchdog/pid#pidfile = /run#' \
-          -e '/\/run\/watchdog\/file/a change = 300'                             \
+          -e "/\/run\/watchdog\/file/a change = $WATCHDOG_CHANGE_TIME"                             \
```
Changing this to more than 300 means we risk getting close to the default test time after an eveimage update (which is 600 seconds). If that is done we might commit to an EVE image which will not work.

rvs replied:
To add to @eriknordmark point -- lets remove the watchdog change from this PR altogether.

It really doesn't belong here (since this one is about cgroups)

Adarsh, once you do that -- I'll be ready to do a final review and merge this

adarsh-zededa replied:
Removed.
